src/Lib.P2P/ConnectionManager.cs

```diff
@@ -0,0 +1,228 @@
+            if (connection.RemotePeer == null)
+                throw new ArgumentNullException("connection.RemotePeer");
+            if (connection.RemotePeer.Id == null)
+                throw new ArgumentNullException("connection.RemotePeer.Id");
+
+            connections.AddOrUpdate(
+                Key(connection.RemotePeer),
+                (key) => new List<PeerConnection> {connection},
+                (key, conns) =>
+                {
+                    if (!conns.Contains(connection)) conns.Add(connection);
+                    return conns;
+                }
+            );
+
+            if (connection.RemotePeer.ConnectedAddress == null)
+                connection.RemotePeer.ConnectedAddress = connection.RemoteAddress;
+            connection.Closed += (s, e) => Remove(e);
+            return connection;
+        }
+
+        /// <summary>
+        ///   Remove a connection.
+        /// </summary>
+        /// <param name="connection">
+        ///   The <see cref="PeerConnection"/> to remove.
+        /// </param>
+        /// <returns>
+        ///   <b>true</b> if the connection was removed; otherwise, <b>false</b>.
+        /// </returns>
+        /// <remarks>
+        ///    The <paramref name="connection"/> is removed from the list of
+        ///    connections and is closed.
+        /// </remarks>
+        public bool Remove(PeerConnection connection)
+        {
+            if (connection == null) return false;
+
+            if (!connections.TryGetValue(Key(connection.RemotePeer), out var originalConns))
+            {
+                connection.Dispose();
+                return false;
+            }
+
+            if (!originalConns.Contains(connection))
+            {
+                connection.Dispose();
+                return false;
+            }
+
+            var newConns = new List<PeerConnection>();
+            newConns.AddRange(originalConns.Where(c => c != connection));
+            connections.TryUpdate(Key(connection.RemotePeer), newConns, originalConns);
+
+            connection.Dispose();
+            if (newConns.Count > 0)
+            {
+                var last = newConns.Last();
+                last.RemotePeer.ConnectedAddress = last.RemoteAddress;
+            }
+            else
+            {
+                connection.RemotePeer.ConnectedAddress = null;
+                PeerDisconnected?.Invoke(this, connection.RemotePeer.Id);
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        ///   Remove and close all connection to the peer ID.
+        /// </summary>
+        /// <param name="id">
+        ///   The ID of a <see cref="Peer"/> to remove.
+        /// </param>
+        /// <returns>
+        ///   <b>true</b> if a connection was removed; otherwise, <b>false</b>.
+        /// </returns>
+        public bool Remove(MultiHash id)
+        {
+            if (!connections.TryRemove(Key(id), out var conns)) return false;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)